Qhull server option

OmarShehata opened the pull request:
Adds a "Use Qhull Server" checkbox. While our built in implementation works, it's slow, and doesn't work for every scenario. This will serve as a backup, and also as a way to check our results (you can click on the checkbox to easily toggle between the two algorithms to see the difference).

The server is running on my personal website, so if that ever goes down, unchecking it will run the built in one, so no worries there.

OmarShehata committed:
Added axios library for making POST requests

OmarShehata committed:
Added QHull mode

OmarShehata committed:
Added Qhull checkbox, default is on

petersbob commented on src/4d.js:
```diff
@@ -323,46 +330,84 @@ var Mode4D = (function (scope) {
+		if(this.useQHull){
+			var params = new URLSearchParams();
+			params.append('points', JSON.stringify(points));
+			axios.post('https://omarshehata.me/qhull/', params)
```
@OmarShehata  A GET request would make more sense since (from what I know) the server is just generating points and giving them back.

OmarShehata replied:
That's a good point! I'll change that.

petersbob commented on src/4d.js:
```diff
@@ -20,6 +20,7 @@ var Mode4D = (function (scope) {
 		this.rightMesh = null;
 
 		this.gridIsVisible = true;
+		this.useQHull = true;
```
@OmarShehata My thoughts is that the QHull server is a fallback solution, not the default, so this would be set to false by default. It would be a good way to also limit unnecessary work for your server (though I'm sure it can handle it).

OmarShehata replied:
I agree, it _should_ be a fallback. The only reason I had it as the default right now was because technically not having it on produced incorrect results. 

I could just have it off by default and let Humke know when he's teaching to turn it on for more accurate/faster results.

petersbob replied:
how are the results technically incorrect when run on the client?

OmarShehata replied:
Sorry, I should have posted this in the original description. So here's a comparison of Qhull vs client algorithm. Notice how it's missing a facet on this tesseract (we generally have an issue with coplanar points in the algorithm - if you wiggle just one of the points there it'll fix it)

![explanation](https://i.imgur.com/kSNvBGJ.gif)

petersbob reviewed (commented):
@OmarShehata Looks pretty good and is working well from what I can tell. What are your thoughts on publishing the server you are running? It would be worth publishing so other people can run it if your server every shuts down. If I remember correctly, isn't it a node server that runs qhull?

OmarShehata commented:
@petersbob re-publishing the server - that would be nice, I was going to say we could even host it on glitch (https://glitch.com/) so it's open source and anyone can fork it, but it requires having you to install Qhull on the server separately so I don't think that'll work. But might be something I can try to put together as a separate repo sometime in the future.

petersbob commented:
@OmarShehata hosting it on glitch sounds like a good idea. I'm all for it.

OmarShehata committed:
Changed to GET request for Qhull server

OmarShehata commented:
I changed the call to be a GET as opposed to a POST as you suggested. I updated my server and just tested it. I think I'm going to keep the qhull checkbox on by default now so that we have it in a presentable state. Thanks for reviewing this @petersbob!
